### ISSUE TITLE
Organize coregistration metadata

### ISSUE THREAD

**Issue by rhugonnet** [GlacioHack/xdem#573]
Several recent PRs including in particular #530 brought up the fact that we need better organization of coregistration metadata.
Currently, every input/output is saved in the same `Coreg.meta` dictionary, which is getting messy and hard to manipulate. Its current keys are summarized in the `CoregDict` typing class: https://github.com/GlacioHack/xdem/blob/02c20268ecfececc2be75e07597ec66997902743/xdem/coreg/base.py#L1157

Below I'm proposing a structure to try to remedy this, and bring more clarity (and practicality?) on the parameters for both users and developers.

First, we would have a nested structure for the `CoregDict`, which could be for instance:
```python
class CoregDict(TypedDict, total=False):

   # Inputs = parameters
    input_random: InRandomDict
    input_fitorbin: InFitOrBinDict
    input_iterative: InIterativeDict
    input_specific: InSpecificDict

    # Outputs
    output_random: OutRandomDict
    output_fitorbin: OutFitOrBinDict
    output_iterative: OutIterativeDict
    output_affine: OutAffineDict
```
where we'd have for instance:
```python
class InRandomDict(TypedDict, total=False):
    # Subsample size input by user, and final size available from data
    subsample: int | float
    # Random state (for subsampling, but also possibly for some fitting methods)
    random_state: int | np.random.Generator | None

class OutRandomDict(TypedDict, total=False):
   # Final subsample size used, depending on amount of valid data
    subsample_final: int
```
Similarly, the `FitOrBin` dictionaries would have user parameters (`fit_optimizer`, `bin_sizes`, etc) on input side, and outputs (`fit_params`, `bin_df`) on the other side.

The advantage of this for the developing side, despite the added structure, is that we can easily extract a group of parameter saved at different stages (usually initialization) for a specific function.
For example:
```python
class MyNewCoreg:
    def _fit_rst():
         xxx = self._subsampling(yyy)
    def _subsampling():
             inputs_random = self.meta["inputs_random"]  # This contains "subsample" and "random_state"
```
or for plotting in the future:
```python
class Coreg:
   def plot_fit():
        outputs_fitorbin = self.meta["outputs_fitorbin"] # Contains everything necessary to plot the fit/bin
```
The problem with the nested structure would be for users to navigate the nested dictionary... But this should become clear quickly with a simple print (and a nice guide in the documentation):
```python
print({"random": {"subsample": 1000, "random_state": 42}})
```
```
{'random': {'subsample': 1000, 'random_state': 42}}
```
Additionally, we could code a `Coreg.info()` method that summarizes everything nicely in words:
```python
Coreg.info()
```
```
Summary of coregistration metadata
Inputs:
    Randomization:
         Subsample size required by user: 1000
         Random state for randomization: 42
         ...
```

**Comment by rhugonnet** [comment]
@adehecq @erikmannerfelt With your green light, I can start on this soon in the continuation of #530.

**Comment by rhugonnet** [comment]
Update: it seemed important to finalize ongoing changes and publish the new documentation, so I moved ahead and implemented this in #530.

**Comment by rhugonnet** [comment]
As expected, the structure makes it much more organized, and allows to solve typing problems that arose when we extracted individual arguments (see #571).
Additionally, it makes it easy to print a nice `info()` :smile: :
See example:
```python
import xdem
c = xdem.coreg.NuthKaab()
c.info()
```
```
Generic coregistration information 
  Method:       NuthKaab 
  Is affine?    True 
  Fit called?   False 
Inputs
  Randomization
    Subsample size requested:                                500000.0
  Fitting and binning
    Fit, bin or bin+fit:                                     bin_and_fit
    Function to fit:                                         _nuth_kaab_fit_func
    Optimizer for fitting:                                   curve_fit
    Bin sizes or edges:                                      80
    Binning statistic:                                       nanmedian
  Iterative
    Maximum number of iterations:                            10
    Tolerance to reach (pixel size):                         0.05
Outputs
  None yet (fit not called)
```
